packages/unified-selection/src/unified-selection/SelectionScope.ts

```diff
@@ -3,277 +3,217 @@
+async function* computeModelSelection(queryExecutor: IECSqlQueryExecutor, ids: string[]): AsyncIterableIterator<SelectableInstanceKey> {
+  const bindings: ECSqlBinding[] = [];
+  const query = `
+    SELECT DISTINCT m.ECInstanceId, ec_classname(m.ECClassId, 's.c') AS ClassName
+    FROM BisCore.Model m JOIN BisCore.Element e ON e.Model.Id = m.ECInstanceId
+    WHERE ${formIdBindings("e.ECInstanceId", ids, bindings)};`;
 
-    yield* executeQuery(queryExecutor, query, bindings);
-  }
+  yield* executeQuery(queryExecutor, query, bindings);
+}
 
-  function getNearestFunctionalElementQuery(ids: string[], bindings: ECSqlBinding[]): string {
-    return `Element2dNearestFunctionalElement (OriginalECInstanceId, OriginalECClassId, ECInstanceId, ECClassId, ParentId) AS (
-              SELECT ECInstanceId, ECClassId, ECInstanceId, ECClassId, Parent.Id FROM BisCore.GeometricElement2d WHERE ${formIdBindings("ECInstanceId", ids, bindings)}
-              UNION ALL
-              SELECT e.OriginalECInstanceId, e.OriginalECClassId, COALESCE(r.TargetECInstanceId, ge.ECInstanceId), COALESCE(r.TargetECClassId, ge.ECClassId), ge.Parent.Id
-              FROM Element2dNearestFunctionalElement e
-                LEFT JOIN BisCore.Element ge ON ge.ECInstanceId = e.ParentId
-                LEFT JOIN Functional.DrawingGraphicRepresentsFunctionalElement r ON r.SourceECInstanceId = e.ECInstanceId
-              WHERE e.ECClassId IS NOT (Functional.FunctionalElement) AND (e.ParentId IS NOT NULL OR r.TargetECInstanceId IS NOT NULL)
-            ),
-            FoundNearestFunctionalElements (OriginalECInstanceId, ECInstanceId, ECClassId) AS (
-              SELECT OriginalECInstanceId, ECInstanceId, ECClassId
-              FROM Element2dNearestFunctionalElement
-              WHERE ECClassId IS (Functional.FunctionalElement)
-            ),
-            ElementsWithoutFunctionalElement (ECInstanceId, ECClassId) AS (
-              SELECT e.OriginalECInstanceId, OriginalECClassId
-              FROM Element2dNearestFunctionalElement e
-                LEFT JOIN FoundNearestFunctionalElements f ON f.OriginalECInstanceId = e.OriginalECInstanceId
-              WHERE e.ParentId IS NULL AND f.ECInstanceId IS NULL
-            ),
-            MergedElements (ECInstanceId, ECClassId) AS (
-              SELECT ECInstanceId, ECClassId FROM FoundNearestFunctionalElements
-              UNION
-              SELECT ECInstanceId, ECClassId FROM ElementsWithoutFunctionalElement
-            ),`;
-  }
+async function* computeFunctionalElementSelection(
+  queryExecutor: IECSqlQueryExecutor,
+  ids: string[],
+  ancestorLevel: number,
+): AsyncIterableIterator<SelectableInstanceKey> {
+  const bindings: ECSqlBinding[] = [];
+  const recurseUntilRoot = ancestorLevel < 0;
+  const query = `
+    WITH RECURSIVE
+      Non3dElementsOrNearestFunctionalElements (OriginalECInstanceId, OriginalECClassId, ECInstanceId, ECClassId, ParentId) AS (
```

Comment:
Could we rename "non3d" to "2d"?

Reply:
We can as long as we don't start the name with a number, renamed.